File: frontend/src/api/endpoints/analysis.ts
import { apiClient } from '../client'
import type {
  DiscoverAgentsResponse,
  RunAnalysisRequest,
  RunAnalysisResponse,
} from '../../types/models'

export interface GenerateMemoRequest {
  company_name: string
  agent_weights: {
    [key: string]: number // e.g., { "business": 25, "market": 25, ... }
  }
}

export interface GenerateMemoResponse {
  job_id: string
  message: string
  company_name: string
}

export const analysisApi = {
  /**
   * Discover available analysis agents
   */
  discoverAgents: async (): Promise<DiscoverAgentsResponse> => {
    const response = await apiClient.get<DiscoverAgentsResponse>(
      '/v1/analysis/discover-agents'
    )
    return response.data
  },

  /**
   * Run multi-agent analysis with selected agents
   */
  runAnalysis: async (request: RunAnalysisRequest): Promise<RunAnalysisResponse> => {
    const response = await apiClient.post<RunAnalysisResponse>(
      '/v1/analysis/run-agents',
      request
    )
    return response.data
  },

  /**
   * Generate final investment memo with weighted analysis
   */
  generateMemo: async (request: GenerateMemoRequest): Promise<GenerateMemoResponse> => {
    const response = await apiClient.post<GenerateMemoResponse>(
      '/v1/analysis/generate-memo',
      request
    )
    return response.data
  },
}

File: backend/src/api/models/analysis.py
"""
Pydantic models for analysis API endpoints.
"""

from pydantic import BaseModel, Field, validator
from typing import List, Optional, Any, Dict


class RunAnalysisRequest(BaseModel):
    """Request model for running multi-agent analysis"""
    company_name: str = Field(..., description="Company name from Phase 1")
    selected_agents: List[str] = Field(
        ...,
        description="List of agent types to run (e.g., ['business', 'market'])",
        min_length=1
    )


class AgentInfo(BaseModel):
    """Information about an available agent"""
    agent_type: str = Field(..., description="Agent identifier (e.g., 'business', 'market')")
    agent_name: str = Field(..., description="Human-readable name")
    description: str = Field(..., description="Agent description")
    available: bool = Field(..., description="Whether agent is available/discovered")


class DiscoverAgentsResponse(BaseModel):
    """Response model for agent discovery"""
    agents: List[AgentInfo]
    total_agents: int
    available_agents: int


class RunAnalysisResponse(BaseModel):
    """Response model for starting analysis"""
    job_id: str
    message: str
    company_name: str
    selected_agents: List[str]


class SimulateQARequest(BaseModel):
    """Request model for simulating Q&A responses"""
    company_name: str = Field(..., description="Company name from Phase 1")

    class Config:
        json_schema_extra = {
            "example": {
                "company_name": "TechVenture AI"
            }
        }


class GenerateMemoRequest(BaseModel):
    """Request model for generating final investment memo"""
    company_name: str = Field(..., description="Company name")
    agent_weights: Dict[str, int] = Field(
        ...,
        description="Agent weights as key-value pairs (e.g., {'business': 25, 'market': 25, ...})"
    )

    @validator('agent_weights')
    def validate_weights(cls, v):
        """Validate that weights sum to 100"""
        total = sum(v.values())
        if total != 100:
            raise ValueError(f"Agent weights must sum to 100, got {total}")
        return v

    class Config:
        json_schema_extra = {
            "example": {
                "company_name": "TechVenture AI",
                "agent_weights": {
                    "business": 25,
                    "market": 25,
                    "tech": 25,
                    "risk": 25
                }
            }
        }


class GenerateMemoResponse(BaseModel):
    """Response model for memo generation job"""
    job_id: str
    message: str
    company_name: str
